File: frontend/src/lib/knowledge.ts
import { api } from "@/lib/api";

export type KnowledgeCategory = {
  id: number;
  name: string;
  complete_name: string | null;
  article_count: number;
  active: boolean;
};

export type KnowledgeArticle = {
  id: number;
  article_ref: string | null;
  name: string;
  state: string;
  category: [number, string] | null;
  owner_agent: [number, string] | null;
  task: [number, string] | null;
  version: number;
  published_at: string | null;
};

export async function fetchKnowledgeCategories() {
  const response = await api.get<KnowledgeCategory[]>("/api/v1/knowledge/categories");
  return response.data;
}

export async function createKnowledgeCategory(payload: {
  name: string;
  description?: string;
  parent_id?: number;
}) {
  const response = await api.post<KnowledgeCategory>("/api/v1/knowledge/categories", payload);
  return response.data;
}

export async function fetchKnowledgeArticles(state?: string) {
  const response = await api.get<KnowledgeArticle[]>("/api/v1/knowledge/articles", {
    params: state ? { state } : {},
  });
  return response.data;
}

export async function createKnowledgeArticle(payload: {
  name: string;
  summary?: string;
  content_html?: string;
  category_id?: number;
  task_id?: number;
}) {
  const response = await api.post<KnowledgeArticle>("/api/v1/knowledge/articles", payload);
  return response.data;
}

export async function transitionKnowledgeArticle(
  articleId: number,
  action: "submit_review" | "publish" | "archive" | "reset_draft",
) {
  const response = await api.post<KnowledgeArticle>(
    `/api/v1/knowledge/articles/${articleId}/transition`,
    { action },
  );
  return response.data;
}

File: backend/app/schemas/knowledge.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class KnowledgeCategory(BaseModel):
    id: int
    name: str
    complete_name: str | None = None
    active: bool = True
    article_count: int = 0


class KnowledgeCategoryCreate(BaseModel):
    name: str = Field(min_length=2, max_length=160)
    description: str | None = None
    parent_id: int | None = None
    sequence: int = 10
    active: bool = True


class KnowledgeArticle(BaseModel):
    id: int
    article_ref: str | None = None
    name: str
    state: str
    category: tuple[int, str] | None = None
    owner_agent: tuple[int, str] | None = None
    task: tuple[int, str] | None = None
    version: int = 1
    published_at: datetime | None = None


class KnowledgeArticleCreate(BaseModel):
    name: str = Field(min_length=2, max_length=220)
    summary: str | None = None
    content_html: str | None = None
    category_id: int | None = None
    owner_agent_id: int | None = None
    task_id: int | None = None
    tags: str | None = None
    related_ticket_ids: list[int] = Field(default_factory=list)


class KnowledgeArticleUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=2, max_length=220)
    summary: str | None = None
    content_html: str | None = None
    category_id: int | None = None
    owner_agent_id: int | None = None
    task_id: int | None = None
    tags: str | None = None
    related_ticket_ids: list[int] | None = None


class KnowledgeArticleTransition(BaseModel):
    action: Literal["submit_review", "publish", "archive", "reset_draft"]
